tensorflow_graphics/projects/mesh_rcnn/branches/voxel_layer.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2020 The TensorFlow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of the voxel prediction branch of Mesh R-CNN."""
+
+import tensorflow as tf
+from tensorflow.keras import layers as keras_layers
+
+
+class VoxelPredictionLayer(keras_layers.Layer):
+  """Implements the 'Voxel Tube' from Mesh R-CNN.
+
+  Takes image features as input and passes them through a fully convolutional
+  network to keep correspondence between input features and predictions.
+
+  This network produces a feature map with G channels giving a column of voxel
+  occupancy scores for each position in the input.
+  """
+
+  def __init__(self,
+               num_classes,
+               num_convs,
+               latent_dim,
+               out_depth,
+               name='VoxelPredictionLayer'):
+    """Constructs the layer.
+
+    Args:
+      num_classes: `int` scalar denoting the number of different object classes
```

Comment:
You don't have to follow this suggestion for now, but type hints are better done via python3 typing module rather than in doc string. https://docs.python.org/3/library/typing.html

Reply:
Nice hint! I'll change that if i got some time left on the weekend or Monday.